File: server/routes/leaderboard.js
'use strict';

const express = require('express');
const router = express.Router();
const { getDb } = require('../db');

/** GET /api/leaderboard/:videoId?limit=10&offset=0 */
router.get('/:videoId', (req, res) => {
  const db = getDb();
  const { videoId } = req.params;
  const limit = Math.min(parseInt(req.query.limit) || 10, 100);
  const offset = parseInt(req.query.offset) || 0;

  const rows = db.prepare(`
    SELECT username, score, max_combo, combo_rank, created_at
    FROM scores
    WHERE video_id = ?
    ORDER BY score DESC
    LIMIT ? OFFSET ?
  `).all(videoId, limit, offset);

  const total = db.prepare('SELECT COUNT(*) as c FROM scores WHERE video_id = ?').get(videoId).c;

  return res.json({ videoId, total, offset, limit, scores: rows });
});

/** POST /api/leaderboard/:videoId — submit a score */
router.post('/:videoId', (req, res) => {
  const db = getDb();
  const { videoId } = req.params;
  const { userId, score, maxCombo, comboRank } = req.body;

  if (!userId || score === undefined) {
    return res.status(400).json({ error: 'userId and score required' });
  }

  const user = db.prepare('SELECT id, username FROM users WHERE id = ?').get(userId);
  if (!user) return res.status(404).json({ error: 'User not found' });

  db.prepare(`
    INSERT INTO scores (user_id, username, video_id, score, max_combo, combo_rank)
    VALUES (?, ?, ?, ?, ?, ?)
  `).run(user.id, user.username, videoId, score, maxCombo || 0, comboRank || 'E');

  // Return player's rank
  const rank = db.prepare(`
    SELECT COUNT(*) + 1 as rank FROM scores
    WHERE video_id = ? AND score > ?
  `).get(videoId, score).rank;

  return res.json({ success: true, rank });
});

module.exports = router;

File: server/routes/leaderboard.py
from flask import Blueprint, jsonify, request

from db import get_db
from limiter import limiter

leaderboard_bp = Blueprint('leaderboard', __name__)


@leaderboard_bp.get('/<video_id>')
@limiter.limit('120 per minute')
def get_leaderboard(video_id):
    """GET /api/leaderboard/<videoId>?limit=10&offset=0"""
    conn = get_db()
    try:
        limit = min(int(request.args.get('limit', 10)), 100)
    except (ValueError, TypeError):
        limit = 10
    try:
        offset = int(request.args.get('offset', 0))
    except (ValueError, TypeError):
        offset = 0

    rows = conn.execute(
        '''SELECT username, score, max_combo, combo_rank, created_at
           FROM scores
           WHERE video_id = ?
           ORDER BY score DESC
           LIMIT ? OFFSET ?''',
        (video_id, limit, offset)
    ).fetchall()

    total = conn.execute(
        'SELECT COUNT(*) as c FROM scores WHERE video_id = ?', (video_id,)
    ).fetchone()['c']

    return jsonify({'videoId': video_id, 'total': total, 'offset': offset,
                    'limit': limit, 'scores': rows})


@leaderboard_bp.post('/<video_id>')
@limiter.limit('20 per minute')
def submit_score(video_id):
    """POST /api/leaderboard/<videoId> — submit a score."""
    conn = get_db()
    data = request.get_json(silent=True) or {}
    user_id = data.get('userId')
    score = data.get('score')
    max_combo = data.get('maxCombo', 0)
    combo_rank = data.get('comboRank', 'E')

    if not user_id or score is None:
        return jsonify({'error': 'userId and score required'}), 400

    user = conn.execute(
        'SELECT id, username FROM users WHERE id = ?', (user_id,)
    ).fetchone()
    if not user:
        return jsonify({'error': 'User not found'}), 404

    conn.execute(
        '''INSERT INTO scores (user_id, username, video_id, score, max_combo, combo_rank)
           VALUES (?, ?, ?, ?, ?, ?)''',
        (user['id'], user['username'], video_id, score, max_combo, combo_rank)
    )
    conn.commit()

    rank = conn.execute(
        'SELECT COUNT(*) + 1 as rank FROM scores WHERE video_id = ? AND score > ?',
        (video_id, score)
    ).fetchone()['rank']

    return jsonify({'success': True, 'rank': rank})
